src/lib/party.ts

```diff
@@ -516,8 +575,105 @@ class Party {
     return new (Party as any)(body, authorization, party.isSelf);
   }
 
+  token(tokenId?: Field) {
+    let thisParty = this;
+    let customToken = new Token({
+      tokenOwner: thisParty.body.publicKey,
+      parentTokenId: tokenId,
+    });
+    return {
+      id: customToken.id,
+      parentTokenId: customToken.parentTokenId,
+      tokenOwner: customToken.tokenOwner,
+
+      mint({ address, amount }: MintOrBurnParams) {
+        let receiverParty = Party.createUnsigned(address, {
+          caller: this.id,
+          tokenId: this.id,
+          callDepth: 1,
+          useFullCommitment: Bool(true),
+        });
+
+        // Add the amount to mint to the receiver's account
+        receiverParty.body.balanceChange =
+          receiverParty.body.balanceChange.add(amount);
+      },
+
+      burn({ address, amount }: MintOrBurnParams) {
+        let receiverParty = Party.createUnsigned(address, {
+          caller: this.id,
+          tokenId: this.id,
+          callDepth: 1,
+          useFullCommitment: Bool(true),
+        });
+
+        // Sub the amount to burn from the receiver's account
+        receiverParty.body.balanceChange =
+          receiverParty.body.balanceChange.sub(amount);
+
+        // Require signature from the receiver account being deducted
+        receiverParty.authorization = {
+          kind: 'lazy-signature',
+        };
+      },
+
+      send({ from, to, amount }: SendParams) {
+        // Check if the sender party is the current zkApp address
+        if (from === thisParty.publicKey) {
+          // If so, create a new party for the zkApp to send the amount to the receiver
+          let senderParty = Party.createUnsigned(thisParty.publicKey, {
```

Comment:
Actually, I think in the `if` branch, we also have to sign the senderParty. It's a separate party, not connected to whatever authorization the `thisParty` already has (it's even for a different tokenId, so a different account). It also has to be a child of `thisParty` according to the token rules (AFAIU). So, it seems the logic of the `if` branch is actually equivalent to the `else` branch and we can just remove the `if` branch

Reply:
You're right, fixed!